src/rewriter/rewrite_db_proof_cons.cpp

```diff
@@ -41,8 +70,950 @@ bool RewriteDbProofCons::prove(CDProof* cdp,
+        return false;
+      }
+      // the missing transitivity link is a subgoal to prove
+      transEq = stgt.eqNode(target[1]);
+      vcs.push_back(transEq);
+      Trace("rpc-debug2") << "  Try transitive with " << transEq << std::endl;
+    }
+    // do its conditions hold?
+    // Get the conditions, substituted { vars -> subs } and with side conditions
+    // evaluated.
+    if (!rpr.getObligations(vars, subs, vcs))
+    {
+      // cannot get conditions, likely due to failed side condition
+      Trace("rpc-debug2") << "...fail (obligations)" << std::endl;
+      return false;
+    }
+  }
+  // First, check which premises are non-trivial, and if there is a trivial
+  // failure. Those that are non-trivial are added to condToProve.
+  std::vector<Node> condToProve;
+  for (const Node& cond : vcs)
+  {
+    Assert(cond.getKind() == Kind::EQUAL);
+    // substitute to get the condition-to-prove
+    DslProofRule cid;
+    // check whether condition is already known to hold or not hold
+    if (proveInternalBase(cond, cid))
+    {
+      if (cid == DslProofRule::FAIL)
+      {
+        // does not hold, we fail
+        Trace("rpc-debug2") << "...fail (simple condition failure for " << cond
+                            << ")" << std::endl;
+        return false;
+      }
+      // already holds, continue
+      continue;
+    }
+    if (!doRecurse)
+    {
+      // we can't apply recursion, return false
+      Trace("rpc-debug2") << "...fail (recursion limit)" << std::endl;
+      return false;
+    }
+    // save, to check below
+    condToProve.push_back(cond);
+  }
+  // if no trivial failures, go back and try to recursively prove
+  for (const Node& cond : condToProve)
+  {
+    Trace("rpc-infer-sc") << "Check condition: " << cond << std::endl;
+    // recursively check if the condition holds
+    DslProofRule cid = proveInternal(cond);
+    if (cid == DslProofRule::FAIL)
+    {
+      // print reason for failure
+      Trace("rpc-infer-debug")
+          << "required: " << cond << " for " << id << std::endl;
+      return false;
+    }
+  }
+  // successfully found instance of rule
+  if (TraceIsOn("rpc-infer"))
+  {
+    Trace("rpc-infer") << "INFER " << target << " by " << id
+                       << (transEq.isNull() ? "" : " + TRANS") << std::endl;
+  }
+  // cache the success
+  ProvenInfo* pi = &d_pcache[target];
+  if (!transEq.isNull())
+  {
+    Trace("rpc-debug2") << "..." << target << " proved by TRANS" << std::endl;
+    Node transEqStart = target[0].eqNode(transEq[0]);
+    // proves both
+    pi->d_id = DslProofRule::TRANS;
+    pi->d_vars.push_back(transEqStart);
+    pi->d_vars.push_back(transEq);
+    Trace("rpc-debug2") << "...original equality was " << transEqStart
+                        << std::endl;
+    // we also prove the original
+    pi = &d_pcache[transEqStart];
+  }
+  pi->d_id = pic.d_id;
+  if (pic.isInternalRule())
+  {
+    pi->d_vars = pic.d_vars;
```

Comment:
Maybe use std::move() here instead of copying the vector, since `pic.d_vars` will not be used anymore afterwards.